### ISSUE TITLE
Unexpected parser error

### ISSUE THREAD

**Issue by rodrigues** [rrrene/credo#729]
### Environment

* Credo version (`mix credo -v`): `1.1.5`
* Erlang/Elixir version (`elixir -v`):

```
Erlang/OTP 22 [erts-10.5.6] [source] [64-bit] [smp:4:4] [ds:4:4:10] [async-threads:1] [hipe]

Elixir 1.9.4 (compiled with Erlang/OTP 20)
```

* Operating system: macOS Catalina

### What were you trying to do?

For one codebase I've got a parser error, and reproduced the error in this repo: https://github.com/rodrigues/credo_interpolation_error

```elixir
defmodule CredoInterpolationError do
  @moduledoc ~S"""
  Example to demonstrate a problem with credo 1.1.5 and:
  - elixir 1.9.4
  - otp 22.1.8
  """

  def foo do
    env = Application.get_env(:credo_interpolation_error, :foo)

    case "#{env}" do
      "A" <> _ = env -> "🇿🇼 #{String.upcase(env)}"
      "B" <> _ = env -> "🇻🇺 #{String.upcase(env)}"
      env -> String.upcase(env)
    end
  end
end
```

### Expected outcome

No issues.

### Actual outcome

For some reason this seems to happen when there are flag emojis in the mix. Tried other emojis and couldn't reproduce the error, but it reproduced with other flags 🤔

There's also no actual extra whitespace, when I change to another emoji the code readability warning is also not there anymore.

```
➜ mc
Checking 1 source file ...

info: some checks were skipped because they're not compatible with
your version of Elixir (1.9.4).

You can deactivate these checks by adding this to the `checks` list in your config:

    {Credo.Check.Refactor.MapInto, false},
    {Credo.Check.Warning.LazyLogging, false},

warning: Unexpected error while parsing lib/credo_interpolation_error.ex: [%Credo.Issue{category: :error, check: Credo.Code.ParserError, column: nil, exit_status: 0, filename: "nofilename", line_no: 13, message: "unexpected token: ", meta: [], priority: 0, scope: nil, severity: nil, trigger: nil}]
  lib/credo/check/readability/string_sigils.ex:50: Credo.Check.Readability.StringSigils.run/2
  lib/credo/check/runner.ex:168: Credo.Check.Runner.do_run_check_on_single_source_file/3
  lib/credo/check/runner.ex:161: Credo.Check.Runner.run_check_on_single_source_file/3
  (elixir) lib/enum.ex:1336: Enum."-map/2-lists^map/1-0-"/2
  (elixir) lib/enum.ex:1336: Enum."-map/2-lists^map/1-0-"/2


  Code Readability
┃
┃ [R] ↘ There should be no trailing white-space at the end of a line.
┃       lib/credo_interpolation_error.ex:13:36 #(CredoInterpolationError.foo)

Please report incorrect results: https://github.com/rrrene/credo/issues

Analysis took 0.06 seconds (0.00s to load, 0.06s running 52 checks)
2 mods/funs, found 1 code readability issue.
```

**Comment by philipdexter** [comment]
Some input: I can't reproduce, but I'm on linux and with erts 10.6.

**Comment by rodrigues** [comment]
@philipdexter weird, I could still reproduce the error in OTP 22.2.1 (erts 10.6.1, elixir 1.9.4)

I've also reduced [the code](https://github.com/rodrigues/credo_interpolation_error) to a minimum that reproduces the problem:

```elixir
defmodule CredoInterpolationError do
  def foo(arg) do
    "🇿🇼 #{arg}"
  end
end
```

**Comment by philipdexter** [comment]
@rodrigues here's what I get. Hopefully someone else with linux or macos can chime in.

```
philip.stkhlm> hub clone rodrigues/credo_interpolation_error
Cloning into 'credo_interpolation_error'...
remote: Enumerating objects: 22, done.
remote: Counting objects: 100% (22/22), done.
remote: Compressing objects: 100% (18/18), done.
remote: Total 22 (delta 5), reused 20 (delta 3), pack-reused 0
Receiving objects: 100% (22/22), done.
Resolving deltas: 100% (5/5), done.
philip.stkhlm> cd credo_interpolation_error
philip.stkhlm> mix deps.get
Resolving Hex dependencies...
Dependency resolution completed:
Unchanged:
  bunt 0.2.0
  credo 1.1.5
  jason 1.1.2
* Getting credo (Hex package)
* Getting bunt (Hex package)
* Getting jason (Hex package)
philip.stkhlm> mix credo
==> bunt
Compiling 2 files (.ex)
Generated bunt app
==> jason
Compiling 8 files (.ex)
Generated jason app
==> credo
Compiling 193 files (.ex)
Generated credo app
Checking 1 source file ...

info: some checks were skipped because they're not compatible with
your version of Elixir (1.9.4).

You can deactivate these checks by adding this to the `checks` list in your config:

    {Credo.Check.Refactor.MapInto, false},
    {Credo.Check.Warning.LazyLogging, false},


Please report incorrect results: https://github.com/rrrene/credo/issues

Analysis took 0.00 seconds (0.00s to load, 0.00s running 45 checks)
2 mods/funs, found no issues.

Use `--strict` to show all issues, `--help` for options.
```

**Comment by rodrigues** [comment]
@philipdexter voilà, probably figured out why you don't get the error:

I run my super alias `mc` 🤘, that means `mix credo --all --strict`. Without the `--strict` I also don't get the issue, but `--strict` triggers it, likely for you too?

**Comment by philipdexter** [comment]
@rodrigues yup, I can confirm now that with `--strict` I see the same error.

**Comment by philipdexter** [comment]
Seems like there's a bug in [`Credo.Code.InterpolationHelper.replace_interpolations`](https://github.com/rrrene/credo/blob/master/lib/credo/code/interpolation_helper.ex#L7)

Note the missing closing quotation (and leftover `#`) for source but not for source2.
```elixir
iex(27)> source
"defmodule CredoInterpolationError do\n  def foo(arg) do\n    \"🇿🇼 \#{arg}\"\n  end\nend\n"
iex(28)> InterpolationHelper.replace_interpolations(source)
"defmodule CredoInterpolationError do\n  def foo(arg) do\n    \"🇿🇼 #      \n  end\nend\n"
iex(33)> source2 = "defmodule CredoInterpolationError do\n  def foo(arg) do\n    \"\#{arg}\"\n  end\nend\n"
"defmodule CredoInterpolationError do\n  def foo(arg) do\n    \"\#{arg}\"\n  end\nend\n"
iex(34)> InterpolationHelper.replace_interpolations(source2)
"defmodule CredoInterpolationError do\n  def foo(arg) do\n    \"      \"\n  end\nend\n"

```

**Comment by philipdexter** [comment]
Actually this seems like an issue with `:elixir_tokenizer.tokenize`.

Note the different column numbers for the strings.
```elixir
iex(31)> source
"defmodule CredoInterpolationError do\n  def foo(arg) do\n    \"🇿🇼 \#{arg}\"\n  end\nend\n"
iex(32)> source2
"defmodule CredoInterpolationError do\n  def foo(arg) do\n    \"h \#{arg}\"\n  end\nend\n"
iex(33)> source |> String.to_charlist() |> :elixir_tokenizer.tokenize(1, file: "blah.ex")
{:ok,
 [
   {:identifier, {1, 1, nil}, :defmodule},
   {:alias, {1, 11, nil}, :CredoInterpolationError},
   {:do, {1, 35, nil}},
   {:eol, {1, 37, 1}},
   {:identifier, {2, 3, nil}, :def},
   {:paren_identifier, {2, 7, nil}, :foo},
   {:"(", {2, 10, nil}},
   {:identifier, {2, 11, nil}, :arg},
   {:")", {2, 14, nil}},
   {:do, {2, 16, nil}},
   {:eol, {2, 18, 1}},
   {:bin_string, {3, 5, nil},
    [
      "🇿🇼 ",
      {{3, 9, nil}, {3, 14, nil}, [{:identifier, {3, 11, nil}, :arg}]}
    ]},
   {:eol, {3, 16, 1}},
   {:end, {4, 3, nil}},
   {:eol, {4, 6, 1}},
   {:end, {5, 1, nil}},
   {:eol, {5, 4, 1}}
 ]}
iex(34)> source2 |> String.to_charlist() |> :elixir_tokenizer.tokenize(1, file: "blah.ex")
{:ok,
 [
   {:identifier, {1, 1, nil}, :defmodule},
   {:alias, {1, 11, nil}, :CredoInterpolationError},
   {:do, {1, 35, nil}},
   {:eol, {1, 37, 1}},
   {:identifier, {2, 3, nil}, :def},
   {:paren_identifier, {2, 7, nil}, :foo},
   {:"(", {2, 10, nil}},
   {:identifier, {2, 11, nil}, :arg},
   {:")", {2, 14, nil}},
   {:do, {2, 16, nil}},
   {:eol, {2, 18, 1}},
   {:bin_string, {3, 5, nil},
    ["h ", {{3, 8, nil}, {3, 13, nil}, [{:identifier, {3, 10, nil}, :arg}]}]},
   {:eol, {3, 15, 1}},
   {:end, {4, 3, nil}},
   {:eol, {4, 6, 1}},
   {:end, {5, 1, nil}},
   {:eol, {5, 4, 1}}
 ]}
```

My guess is that this issue affects anything with more than 1 code point. For example, the pirate flag has 4 code points and messes up in an even bigger way:

```elixir
iex(38)> source3
"defmodule CredoInterpolationError do\n  def foo(arg) do\n    \"🏴‍☠️ \#{arg}\"\n  end\nend\n"
iex(39)> InterpolationHelper.replace_interpolations(source3)
"defmodule CredoInterpolationError do\n  def foo(arg) do\n    \"🏴‍☠️ \#{a      \n  end\nend\n"

```

**Comment by rrrene** [comment]
We've had problems with `:elixir_tokenizer.tokenize` in the past. This is the first time with Elixir `1.9`. 

This is certainly a problem. I am not sure if we could build a workaround for this by removing those UTF8 characters (is character the correct term here?) before we run the tokenizer on them. Ideas?

**Comment by philipdexter** [comment]
Perhaps modify the output of `tokenize` by detecting these situations. Or, I suppose something quick like
```elixir
iex(1)> pirate_flag = "🏴‍☠️"
"🏴‍☠️"
iex(2)> "#{pirate_flag} some text" |> String.graphemes() |> Enum.map(&(hd(String
.codepoints(&1))))
["🏴", " ", "s", "o", "m", "e", " ", "t", "e", "x", "t"]
iex(3)> Enum.join(v(-1), "")
"🏴 some text"
```
could work. I guess it depends what we want the end result to be. In this case we probably don't care about the contents of the string, so this could work here.

**Comment by rrrene** [comment]
Thx for reporting this! :+1: It should be be fixed in the latest RC of Credo (`v1.2.0-rc4`).

If it is not, please feel free to re-open this issue!

**Comment by rodrigues** [comment]
@rrrene I've tested and it works fine now, thank you!
